### PR TITLE
Add function learning eval + move docker_exec to shared evals

### PR DESCRIPTION
**Pull request by agentydragon**
## Summary

- New eval: model learns a secret boolean function f: {0,1}^8 → {0,1}^4 by querying inputs and submitting Python program guesses
- Each turn uses `play_turn(query, program)` — scaffold evaluates program against all 256 inputs in isolated Docker exec calls (parallelized, one per input)
- Score = total Hamming loss summed over all turns (lower is better, no LLM judge)
- Three hand-coded functions: `linear_simple` (GF(2) matrix), `junta_3` (3-bit dependency), `parity_groups` (XOR pairs)
- Moves `docker_exec.py` from `twenty_questions/x/shared/` to shared `evals/` directory for cross-eval reuse

## Test plan

- [x] `bazel test //skills/info_gathering/evals/function_learning:test_functions` — function properties
- [x] `bazel test //skills/info_gathering/evals/function_learning:test_function_learning` — game loop with replay client
- [x] `bazel build //skills/info_gathering/evals/function_learning/...` — all 8 targets build
- [x] 20Q autogen binary still builds after docker_exec move
- [ ] Manual live test with `--turn-limit 3` for quick smoke test

https://claude.ai/code/session_01Ez2vG2NbTdDhGKCaQ2Udj6

### CODE REVIEW

File: skills/info_gathering/evals/function_learning/function_learning.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+            f"(eval {scoring.total_eval_s:.1f}s, {scoring.mean_per_input_s * 1000:.0f}ms/input)",
+        )
+        logger.info(
+            "Turn %d: hamming_loss=%d (eval %.1fs, %.0fms/input avg, %.0fms/input max)",
+            game.turn,
+            scoring.hamming_loss,
+            scoring.total_eval_s,
+            scoring.mean_per_input_s * 1000,
+            scoring.max_per_input_s * 1000,
+        )
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply skills/info_gathering/evals/function_learning/function_learning.py:112]: Fix it @copilot

File: skills/info_gathering/evals/function_learning/function_learning.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+            "hamming_loss": scoring.hamming_loss,
+            "total_possible_loss": 2**secret_fn.n * secret_fn.m,
+        }
+        if scoring.errors:
+            response["errors"] = [{"input": e.input, "error": e.error} for e in scoring.errors]
+            response["total_errors"] = len(scoring.errors)
+
+        return json.dumps(response, indent=2)
+
+    return FunctionTool(
```
[the remark this answers is not in the document]
Comment (agentydragon) [reply skills/info_gathering/evals/function_learning/function_learning.py:130]: @copilot fix it, remove the extra param

File: skills/info_gathering/evals/function_learning/scoring.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        raw = "\n".join(block.text for block in mcp_result.content if isinstance(block, TextContent))
+        try:
+            for raw_line in raw.splitlines():
+                stripped = raw_line.strip()
+                if stripped.startswith("{"):
+                    data = json.loads(stripped)
+                    exit_info = data.get("exit", {})
+                    stdout = data.get("stdout", "").strip()
+                    stderr = data.get("stderr", "").strip()
+                    if exit_info.get("kind") == "timed_out":
```
Comment (agentydragon) [inline skills/info_gathering/evals/function_learning/scoring.py]: Wow this is the fucking ugliest. We have pydantic types, use them @copilot 

Properly. And fix that problem with not handling truncated output rv

File: skills/info_gathering/evals/function_learning/scoring.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                        msg = f"Wrong length: expected {secret_fn.m} chars, got {len(r.output)} ({r.output[:50]!r})"
+                    else:
+                        msg = f"Non-binary characters in output: {r.output!r}"
+                    program_errors.append(ProgramError(input=r.input_bits, error=msg))
+            else:
+                hamming_loss += _hamming_distance(expected, r.output)
+        else:
+            hamming_loss += secret_fn.m
+            if len(program_errors) < _MAX_REPORTED_ERRORS:
+                program_errors.append(ProgramError(input=r.input_bits, error="No output"))
```
Comment (agentydragon) [inline skills/info_gathering/evals/function_learning/scoring.py]: You should really pull out a per input function then aggregate over its results @copilot

File: skills/info_gathering/evals/function_learning/test_functions.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    assert PARITY_GROUPS.evaluate("10000000") == "1000"
+    assert PARITY_GROUPS.evaluate("01000000") == "1000"
+    assert PARITY_GROUPS.evaluate("00110000") == "0000"
+    assert PARITY_GROUPS.evaluate("00100000") == "0100"
+    assert PARITY_GROUPS.evaluate("10101010") == "1111"
+    assert PARITY_GROUPS.evaluate("11111111") == "0000"
+
+
+def test_all_functions_have_correct_dimensions() -> None:
+    for fn in [LINEAR_SIMPLE, JUNTA_3, PARITY_GROUPS]:
```
Comment (agentydragon) [inline skills/info_gathering/evals/function_learning/test_functions.py:70]: Parameterized test

Comment (agentydragon) [reply skills/info_gathering/evals/function_learning/test_functions.py:70]: @copilot
